src/storage/sql_table.cpp

```diff
@@ -1,20 +1,314 @@
 #include "storage/sql_table.h"
 
+#include <parser/expression/constant_value_expression.h>
 #include <map>
-#include <set>
 #include <string>
 #include <vector>
 
 #include "common/macros.h"
 #include "storage/storage_util.h"
+#include "type/transient_value_peeker.h"
+#include "type/type_util.h"
+#include "util/time_util.h"
 
 namespace terrier::storage {
 
 SqlTable::SqlTable(const common::ManagedPointer<BlockStore> store, const catalog::Schema &schema)
     : block_store_(store) {
+  // Initialize a DataTable
+  tables_ = {{layout_version_t(0),
+              CreateTable(common::ManagedPointer<const catalog::Schema>(&schema), store, layout_version_t(0))}};
+}
+
+bool SqlTable::Select(const common::ManagedPointer<transaction::TransactionContext> txn, const TupleSlot slot,
+                      ProjectedRow *const out_buffer, layout_version_t layout_version) const {
+  // get the version of current tuple slot
+  const auto tuple_version = slot.GetBlock()->data_table_->layout_version_;
+
+  TERRIER_ASSERT(tuple_version <= layout_version,
+                 "The iterator should not go to data tables with more recent version than the current transaction.");
+
+  if (tuple_version == layout_version) {
+    // when current version is same as the intended layout version, get the tuple without transformation
+    return tables_.at(tuple_version).data_table_->Select(txn, slot, out_buffer);
+  }
+
+  // the tuple exists in an older version.
+  // TODO(schema-change): handle versions from add and/or drop column only
+  col_id_t ori_header[out_buffer->NumColumns()];
+
+  auto desired_v = tables_.at(layout_version);
+  auto tuple_v = tables_.at(tuple_version);
+  auto missing UNUSED_ATTRIBUTE = AlignHeaderToVersion(out_buffer, tuple_v, desired_v, &ori_header[0]);
+  auto result = tables_.at(tuple_version).data_table_->Select(txn, slot, out_buffer);
+
+  // copy back the original header
+  std::memcpy(out_buffer->ColumnIds(), ori_header, sizeof(col_id_t) * out_buffer->NumColumns());
+
+  // fill in missing columns and default values
+  FillMissingColumns(out_buffer, desired_v);
+  return result;
+}
+
+std::pair<bool, TupleSlot> SqlTable::Update(const common::ManagedPointer<transaction::TransactionContext> txn,
```

Comment:
Maybe take in a pointer to a TupleSlot rather than returning a pair? Should be cleaner code, since a lot of times you are just using the bool.